util/auto_tune_compressor.cc

```diff
@@ -144,138 +153,211 @@ CostAwareCompressor::CostAwareCompressor(const CompressionOptions& opts)
-      wa->cost_predictors_.emplace_back();
-      continue;
-    } else {
-      std::vector<IOCPUCostPredictor*> predictors_diff_levels;
-      predictors_diff_levels.reserve(kCompressionLevels[type - 1].size());
-      for (size_t j = 0; j < kCompressionLevels[type - 1].size(); j++) {
-        predictors_diff_levels.emplace_back(new IOCPUCostPredictor(10));
-      }
-      wa->cost_predictors_.emplace_back(std::move(predictors_diff_levels));
+  wa->cost_predictors_.reserve(compressors_.size());
+  for (size_t i = 0; i < kCompressionLevels.size(); i++) {
+    for (size_t j = 0; j < kCompressionLevels[i].size(); j++) {
+      wa->cost_predictors_.emplace_back(new IOCPUCostPredictor(kWindow));
     }
   }
   return ManagedWorkingArea(wa, this);
 }
-void CostAwareCompressor::ReleaseWorkingArea(WorkingArea* wa) {
+void AutoCompressionAlgoLevelSelector::MeasureUtilization() {
+  usage_tracker_.Record();
+}
+void AutoCompressionAlgoLevelSelector::ReleaseWorkingArea(WorkingArea* wa) {
   // remove all created cost predictors
-  for (auto& prdictors_diff_levels :
+  for (auto& predictor :
        static_cast<CostAwareWorkingArea*>(wa)->cost_predictors_) {
-    for (auto& predictor : prdictors_diff_levels) {
-      delete predictor;
-    }
+    delete predictor;
   }
   delete static_cast<CostAwareWorkingArea*>(wa);
 }
+size_t AutoCompressionAlgoLevelSelector::SelectCompressionBasedOnGoal(
+    CostAwareWorkingArea* wa) {
+  // If no budgets are available, use default choice
+  if (!cpu_budget_ || !io_budget_) {
+    size_t default_choice = 0;
+    return default_choice;
+  }
+  if ((block_count_.fetch_add(1, std::memory_order_relaxed)) %
+          kDecideEveryNBlocks !=
+      0) {
+    return cur_compressor_idx_;
+  }
+  MeasureUtilization();
+  auto cpu_util = usage_tracker_.GetCpuUtilization();
+  auto io_util = usage_tracker_.GetIoUtilization();
+  // Select compression whose cpu cost and io cost are within budget
+  // Return the first compression type and level that fits within budget
+  // Get available budgets
+  auto cpu_goal = cpu_budget_->GetRate() / kMicrosInSecond;
+  auto io_goal = io_budget_->GetRate();
+  static constexpr double kAcceptableLowCpuGoal = 0.8;
+  static constexpr double kAcceptableLowIOGoal = 0.9;
+  // Detect the 4 quadrant that we want to explore
+  // Stable region is between the (kAcceptableLowCpuGoal * cpu_goal) to cpu_goal
+  // and  and (kAcceptableLowIOGoal * io_goal) to io_goal
+  bool increase_io = io_util < (kAcceptableLowIOGoal * io_goal);
+  bool decrease_io = io_util > io_goal;
+  bool increase_cpu = cpu_util < (kAcceptableLowCpuGoal * cpu_goal);
+  bool decrease_cpu = cpu_util > cpu_goal;
+  // If we are in the stable region, then we just go ahead with the current
+  // compression type and level
+  if (!increase_io && !increase_cpu && !decrease_io && !decrease_cpu) {
+    return cur_compressor_idx_;
+  } else if (cur_compressor_idx_ >= compressors_.size()) {
```

Comment:
why would this even happen? can we always keep cur_compressor_idx_ within the bound?